### PR TITLE
Implement Espresso contrib mobile commands

### PR DESCRIPTION
**Pull request by dpgraham**
Mobile commands: `setDate`, `setTime`, `navigateTo`, `scrollToPage`, `openDrawer`, `closeDrawer`

(connect with: https://github.com/appium/appium-espresso-driver/issues/323)

### CODE REVIEW

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/model/SetTimeParams.java [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+@SuppressWarnings("unused")
+public class SetTimeParams extends AppiumParams {
+    private Integer hours;
+    private Integer minutes;
+
+    public Integer getHours() {
+        return hours;
+    }
+
+    public void setHour(Integer hour) {
```
Comment (mykola-mokhnach) [inline espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/model/SetTimeParams.java]: hours?

File: lib/commands/general.js [hunk 2]
```diff
@@ -42,6 +42,93 @@ commands.mobileIsToastVisible = async function (opts = {}) {
...
+  return await this.espresso.jwproxy.command(`/appium/execute_mobile/${util.unwrapElement(element)}/set_time`, 'POST', {
+    hours,
+    minutes,
+  });
+};
+
+commands.mobileNavigateTo = async function (opts = {}) {
+  let {element, menuItemId} = assertRequiredOptions(opts, ['menuItemId', 'element']);
+
+  let menuItemIdAsNumber = parseInt(menuItemId, 0);
```
Comment (mykola-mokhnach) [inline lib/commands/general.js]: , 10

File: lib/commands/general.js [hunk 3]
```diff
@@ -42,6 +42,93 @@ commands.mobileIsToastVisible = async function (opts = {}) {
...
+  let {element, scrollTo, scrollToPage, smoothScroll} = assertRequiredOptions(opts, ['element']);
+
+  // Validate that one of scrollTo or scrollToPage was set
+  if (!util.hasValue(scrollTo) && !util.hasValue(scrollToPage)) {
+    logger.errorAndThrow(`Must set either 'scrollTo' or 'scrollToPage' parameter`);
+  }
+
+  // Validate that only one of scrollTo or scrollToPage was set
+  if (util.hasValue(scrollTo) && util.hasValue(scrollToPage)) {
+    logger.errorAndThrow(`'scrollTo' and 'scrollToPage' are mutually exclusive parameters and cannot be set at the same time`);
```
Comment (mykola-mokhnach) [inline lib/commands/general.js]: I'd rather print a warning here

File: lib/commands/general.js [hunk 4]
```diff
@@ -42,6 +42,93 @@ commands.mobileIsToastVisible = async function (opts = {}) {
...
+  const scrollToOpts = ['first', 'last', 'left', 'right'];
+  if (util.hasValue(scrollTo)) {
+    if (!_.isString(scrollTo) || !scrollToOpts.includes(scrollTo.toLowerCase())) {
+      logger.errorAndThrow(`'scrollTo' options must be one of ${scrollToOpts}`);
+    }
+    scrollTo = scrollTo.toLowerCase();
+  }
+
+  // Validate that scrollToPage is a non-negative integer
+  if (util.hasValue(scrollToPage)) {
```
Comment (mykola-mokhnach) [inline lib/commands/general.js]: I'd rather move this verification to a separate internal function to avoid duplication

**Comment by dpgraham** [comment]
Fixes pushed. I'm using the validate.js library to validate scrollTo parameters now.

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/handlers/NavigateTo.java [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+    public Void handle(NavigateToParams params) throws AppiumException {
+        ViewInteraction viewInteraction = Element.getViewInteractionById(params.getElementId());
+        Integer menuItemId = params.getMenuItemId();
+        try {
+            viewInteraction.perform(NavigationViewActions.navigateTo(menuItemId));
+        } catch (Exception e) {
+            if (e instanceof EspressoException) {
+                throw new AppiumException(String.format("Could not navigate to menu item %s. Reason: %s", menuItemId, e));
+            }
+        }
```
Comment (mykola-mokhnach) [inline espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/handlers/NavigateTo.java]: throw e

Comment (dpgraham) [reply espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/handlers/NavigateTo.java]: Fixed

File: lib/commands/general.js [hunk 6]
```diff
@@ -42,6 +43,93 @@ commands.mobileIsToastVisible = async function (opts = {}) {
...
+commands.mobileCloseDrawer = async function (opts = {}) {
+  const {element, gravity} = assertRequiredOptions(opts, ['element']);
+
+  return await this.espresso.jwproxy.command(`/appium/execute_mobile/${util.unwrapElement(element)}/close_drawer`, 'POST', {
+    gravity
+  });
+};
+
+commands.mobileSetDate = async function (opts = {}) {
+  const {element, year, monthOfYear, dayOfMonth} = assertRequiredOptions(opts, ['element']);
```
Comment (mykola-mokhnach) [inline lib/commands/general.js]: are year, monthOfYear and dayOfMonth also required?

File: lib/commands/general.js [hunk 7]
```diff
@@ -42,6 +43,93 @@ commands.mobileIsToastVisible = async function (opts = {}) {
...
+
+  return await this.espresso.jwproxy.command(`/appium/execute_mobile/${util.unwrapElement(element)}/set_date`, 'POST', {
+    year,
+    monthOfYear,
+    dayOfMonth,
+  });
+};
+
+commands.mobileSetTime = async function (opts = {}) {
+  const {element, hours, minutes} = assertRequiredOptions(opts, ['element']);
```
Comment (mykola-mokhnach) [inline lib/commands/general.js]: are hours and  minutes also required?

File: lib/commands/general.js (hunk 6 above)
Comment (dpgraham) [reply lib/commands/general.js]: Yep, fixed

File: lib/commands/general.js (hunk 7 above)
Comment (dpgraham) [reply lib/commands/general.js]: Yes.

**Comment by mykola-mokhnach** [comment]
LGTM, only one comment is left (parseInt arg)

**Pull request merged**